### ISSUE TITLE
Is there any way to delete records using Query?

### ISSUE THREAD

**Issue by elgris** [aerospike/aerospike-client-go#8]
Hello guys!
I have a question about using Aerospike client. The problem I'm trying to solve is to delete records with the same primary key and different secondary keys. I can get those records with Query command, but how to remove them with a single query?
Retrieving records and then deleting them one by one is not an option :( Writing LUA scripts is not an option as well.

Thanks for any suggestion!

**Comment by mariaefi29** [comment]
Hi guys!

I did everything what you say, but the udf doesn't delete the records. However, if I delete records with the client on keys (using the same statement), it deletes successfully. I've been stuck with this all day( Do you have any ideas what I am doing wrong?

```
stmt := as.NewStatement(s.namespace, state.String(), hotelIDBin, feedVersionBin)

		err := stmt.SetFilter(as.NewEqualFilter(feedVersionBin, feedVersion(feed, versionToDelete)))
		if err != nil {
			return fmt.Errorf("set filter by feed: %w", err)
		}

		//rec, err := s.client.Query(nil, stmt)
		//if err != nil {
		//	return fmt.Errorf("execute query: %w", err)
		//}
		//
		//for hotel := range rec.Records {
		//	_, err := s.client.Delete(nil, hotel.Key)
		//	if err != nil {
		//		return fmt.Errorf("delete: %w", err)
		//	}
		//}
		task, err := s.client.ExecuteUDF(nil, stmt, "delete", "deleteRecord")
		if err != nil {
			return err
		}
		err = <-task.OnComplete()
		return err
	}
```
The commented code work well, but not the ExecuteUDF, records are still present.

Registering is correct as well. The UDF is exactly the same as you proposed.
```
func (s *Store) registerUDF() error {
	task, err := s.client.RegisterUDF(nil, deleteUDF, "delete.lua", as.LUA)
	if err != nil {
		return err
	}

	return <-task.OnComplete()
}
```

Could you please help?

**Comment by khaf** [comment]
This is a VERY old solution for this problem. If you want to delete all the records in a namespace or set, use the Truncate command. If you want to only delete a Subset and you are using a server that supports Expressions, use `Client.QueryExecute` API, with Expressions and Operations.
Otherwise, then UDFs are the way to go. I need to see your UDF to understand what's going on.

**Comment by mariaefi29** [comment]
My UDF is the same as you proposed) I found the solution first on aerospike discuss forum and stackoverflow and then only here. They say it should work(

```
function deleteRecord(rec)
    aerospike:remove(rec)
end
```

https://discuss.aerospike.com/t/how-to-delete-a-record-based-on-secondary-index/2262

This `aerospike:remove(rec)` is still present in the docs, so it still should work(

**Comment by khaf** [comment]
Do you get back the records you want to delete if you pass that statement to a query?

**Comment by mariaefi29** [comment]
We have server 3.14.0, I guess it's not the way to go with operations(

**Comment by khaf** [comment]
I just tested this code, and it works. You need to check and find out why yours:

```go
package main

import (
	"log"
	"time"

	as "github.com/aerospike/aerospike-client-go"
)

const udfDelete = `function deleteRecord(rec)
   aerospike:remove(rec)                   -- Delete main record, Populate the return status
end`

const (
	ns      = "test"
	set     = "del"
	binName = "bin"

	recordCount = 50
)

func main() {

	log.SetFlags(0)

	policy := as.NewClientPolicy()
	policy.Timeout = 5 * time.Second
	client, err := as.NewClientWithPolicyAndHost(policy, as.NewHost("vmu1804", 3000))
	if err != nil {
		log.Fatalln(err)
	}

	client.DropIndex(nil, ns, set, set+binName)

	idxTask, err := client.CreateIndex(nil, ns, set, set+binName, binName, as.NUMERIC)
	if err != nil {
		log.Fatalln(err)
	}
	defer client.DropIndex(nil, ns, set, set+binName)

	if err := <-idxTask.OnComplete(); err != nil {
		log.Fatalln(err)
	}

	regTask, err := client.RegisterUDF(nil, []byte(udfDelete), "udfDelete.lua", as.LUA)
	if err != nil {
		log.Fatalln(err)
	}

	// wait until UDF is created
	if err := <-regTask.OnComplete(); err != nil {
		log.Fatalln(err)
	}

	for i := 0; i < recordCount; i++ {
		// a new record that is not in the range
		key, _ := as.NewKey(ns, set, i)
		if err := client.PutBins(nil, key, as.NewBin(binName, i)); err != nil {
			log.Fatalln(err)
		}
	}

	//count the existing data
	recordset, err := client.ScanAll(nil, ns, set)
	if err != nil {
		log.Fatalln(err)
	}

	i := 0
	for fullRec := range recordset.Results() {
		if fullRec.Err != nil {
			log.Fatalln(fullRec.Err)
		}
		i++
	}

	if i != recordCount {
		log.Fatalln("Error: Record count in the set (", i, ") does not match", recordCount)
	}

	statement := as.NewStatement(ns, set)
	statement.SetFilter(as.NewRangeFilter(binName, 1, recordCount))
	exTask, err := client.ExecuteUDF(nil, statement, "udfDelete", "deleteRecord")
	if err != nil {
		log.Fatalln(err)
	}

	// wait until UDF is run on all records
	if err := <-exTask.OnComplete(); err != nil {
		log.Fatalln(err)
	}

	// read all data and make sure it is consistent
	recordset, err = client.ScanAll(nil, ns, set)
	if err != nil {
		log.Fatalln(err)
	}

	i = 0
	for fullRec := range recordset.Results() {
		if fullRec.Err != nil {
			log.Fatalln(fullRec.Err)
		}
		i++
	}

	if i != 1 {
		log.Fatalln("Error: Record count in the set (", i, ") does not match", recordCount)
	} else {
		log.Println("Success! All records other than one were removed.")
	}
}
```

**Comment by mariaefi29** [comment]
Dear @khaf,

I have been stuck with this, was playing with your code and couldn't see the difference at all...

And then I found a bug!

I had `for` statement and instead of doing this at end:

```
err = <-task.OnComplete()
	if err != nil {
	   return err
	}
```
I did that: 

```
err = <-task.OnComplete()
return err
```
But that was the for statement!(((( thats why records weren't deleted(((

Oh, a lot of time wasted and especially yours((( 
I am so sorry for bothering you! Thank you so much for your help! I really appreciate it!
